Add delayed preloading for arrays of records

dinahshi opened the pull request:
### Summary

Adds method Relation#load_associations(arr, *args) which accepts an array
of records and args to preload. Useful for when a loaded list is filtered
in ruby to produce an Array and we want to preload the records in this Array.
Effectively exposes `Preloader.preload(records, associations, preload_scope = nil)`
without requiring people to know about a separate Preloader class.

### Todo
- [x] Documentation
- [x] CHANGELOG

cc/ @matthewd

dinahshi committed:
add delayed preloading on arrays of records

rafaelfranca commented:
This is awesome. We have an implementation of this at Shopify. This is the implementation.

```
class AssociationPreloader
  def initialize(records)
    @records = records
  end

  def preload(*associations)
    if @records.is_a?(ActiveRecord::Relation)
      @records.preload(associations)
    else
      ActiveRecord::Associations::Preloader.new.preload(@records, associations.flatten) unless @records.empty?
      @records
    end
  end
end
```

Maybe we should also support passing an relation?

Also can you add documentation?

rafaelfranca commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -78,6 +78,18 @@ def #{method_name}=(value)           # def includes_values=(value)
 
     alias extensions extending_values
 
+    def load_associations(arr, *args)
```
`array` is better than `arr` but I think we should also support relations.

kaspth replied:
Could be `loaded_records_or_relation`.

rafaelfranca commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1180,6 +1192,14 @@ def check_if_method_has_arguments!(method_name, args)
       end
 
+      def check_if_array_contains_base_elements!(records)
+        records.flatten.compact.each do |r|
```
`record` seems to be better than `r`.

rafaelfranca commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -78,6 +78,18 @@ def #{method_name}=(value)           # def includes_values=(value)
 
     alias extensions extending_values
 
+    def load_associations(arr, *args)
```
Can we call the `*args` of `associations`?

dinahshi replied:
Do you mean a variable rename here or something else?

kaspth replied:
It's tough to understand what `args` mean here, yes. As a reader I'm confused by `args.each do |associations|`. So it holds an array of arrays of associations? Or something else?

rafaelfranca commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -677,6 +677,44 @@ def test_loading_with_one_association_with_non_preload
     assert_equal Post.find(1).last_comment, post.last_comment
   end
 
+  def test_load_associations
```
There are some style violations on those tests.

bogdan commented:
We are using similar technique in our project.
We found a major problems with this implementation:

### Sometimes loaded associations. 

Preloader currently determines if the association is preloaded or not by the first record https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader.rb#L180
We have encountered that practically: it results in the double load on the loaded target records. I think we need to fix that part first before exposing the power of Preloader to the audience.


### Offtopic

You may see our implementation just for the interest what is efficient for us:

``` ruby
  def preload_into_association(host_association, *preload_associations)
    # Preloading records in such a weird way
    # Because we want to cache them inside an association
    # but not have a separated scope as a result of #preload call
    preload_associations.each do |preload_association|
      collection = send(host_association)
      unpreloaded = collection.reject {|l| l.association(preload_association).loaded?}
      ActiveRecord::Associations::Preloader.new.preload(unpreloaded, preload_association)
    end
  end
```

Here is how it is used:

``` ruby
Site.has_many :campaigns
Campaign.has_many :assets

site.preload_into_association(:campaigns, :assets)
# which is better than
site.campaigns.preload(:assets) 
# because it generates new relation and site.campaigns.map(&:assets) is not cached...EVER...  
```

You may see that we filter out loaded records with loaded association from collection which resolves the original problem I described. The counter problem of that is that it doesn't allow to pass the nested associations... only an array of associations.

kaspth commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -677,6 +677,44 @@ def test_loading_with_one_association_with_non_preload
+    even_posts, odd_posts = posts.partition.each_with_index{ |p, i| i.even? }
+
+    assert_queries(1) do
+      preloaded_posts = Post.load_associations(even_posts, :comments)
```
Since `load_associations` mutates the posts would it be clearer if we just referenced `even_posts` throughout?

kaspth replied:
I also find the argument order hard to understand and I had to read it several times.

Would it be clearer as `Post.preload_associations(:comments, on: even_posts)`? A `preload_association` alias might also be nice 😊

dinahshi commented:
@rafaelfranca I like the Shopify implementation's straightforward way of handling relations. I'm a little concerned that the method can return different types. When passed a relation, it returns a relation so it can be chained with other query methods like `where`, `order`, etc. But when passed an Array it returns an Array.

dinahshi committed:
Add support for relation

dinahshi committed:
Fix Preloader skipping records when first in list is loaded

dinahshi committed:
Allow preloading of different objects with same record

dinahshi committed:
Add preload_associations alias, documentation

dinahshi commented:
Addressed feedback from @rafaelfranca and @kaspth. aab3b41f3d360c56bcd8e72295863676d4b3ed7a fixes the issue raised in #32140 where it is assumed that an entire list is loaded when the first record in the list is loaded. 8d6e1468b6c126f7cfb96171e417d892f257bbdf fixes preloading different object ids.

Added documentation and updated changelog.

lsylvester commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -78,6 +78,33 @@ def #{method_name}=(value)           # def includes_values=(value)
+        records
+      end
+    end
+    alias_method :preload_associations, :load_associations
```
The method `preload_associations` as already been defined in the latest version  by #31250 so I think that this will cause conflicting behaviour.

In #31250, this method is doing the preloading of associations as part of the loading of the relation, so doesn't take any arguments for the different associations, but does accept the records as the preloading can be done on a subset of the records.

I think that this change would still be good without this alias, so maybe it could be removed? 

Otherwise, maybe these methods could be moved to be class methods on `ActiveRecord::Base` as I am not sure if there is use case for using them on a relation (all the examples in the tests are directly on the  `ActiveRecord::Base` subclasses), but there might be other use cases I am not aware of.

Last option would be to rename the other `preload_associations` method, easiest thing to do would be to prefix with `_` as it is not really part of the public API.

dinahshi replied:
Thanks for pointing this out! I think it'd be nice to make `preload_associations` part of the public API. It doesn't seem like there's a precedent for prefixing private methods with `_` so I'd like to rename the current method to be more specific—maybe `preload_associations_on_records`?

In terms of the method being defined on Base vs Relation, I wanted to keep the logic close to the original `preload` query method which is defined on Relation.

sunnyrjuneja commented:
I'm super excited about this feature, thank you so much @dinahshi and @eileencodes (and everyone else who contributed). Just wanted to confirm, is this is targeting Rails 6?

dinahshi committed:
Rename preload_associations to execute_preloads

dinahshi commented:
After a few months on the back burner, I believe this is ready for review! 😅

As suggested by @matthewd, I've removed the preload_associations alias to follow the precedent of "load" implying that an action occurs immediately, and renamed `preload_associations` (introduced by #31250) to `execute_preload` in order to distinguish that the operation occurs on a relation's previously defined associations.

kaspth commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -78,6 +78,32 @@ def #{method_name}=(value)           # def includes_values=(value)
+        end
+        records
+      end
+    end
```
Still think a `load_association` (singular) alias would be nice for the single association case.

kaspth commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -78,6 +78,32 @@ def #{method_name}=(value)           # def includes_values=(value)
     alias extensions extending_values
 
+    # Allows preloading of associations on a relation or array of records,
+    # in the same way that #preload does. Mutates the +records+ argument.
```
Should `preload` be linked here? Or does RDoc do it automatically via `#` and I just don't remember? 😄

dinahshi replied:
Other methods seem to have it linked automatically through `#` but I'm not sure!

kaspth commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1186,6 +1212,14 @@ def check_if_method_has_arguments!(method_name, args)
+      def check_if_array_contains_valid_elements!(records)
+        records.flatten.compact.each do |record|
+          unless record.is_a? ActiveRecord::Base
+            raise ArgumentError, "The method .load_associations() must be given an array of #{self.klass} elements."
```
Could `self.klass` here just be `klass`?

kaspth replied:
I think we'd just say

```ruby
raise ArgumentError, "`load_associations` must be given…"
```

kaspth commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -1186,6 +1212,14 @@ def check_if_method_has_arguments!(method_name, args)
       end
 
+      def check_if_array_contains_valid_elements!(records)
+        records.flatten.compact.each do |record|
```
How likely are users to pass in nested arrays? And ones where `nil` is part of the array?

Does the preloader do the same work on the `records` or does it support nested arrays and nil values?

Since we mutate records how can users properly loop through them if they're either nested or contain nil? E.g. should we punt the `flatten.compact` to user code instead of baking it in here?

Come to think of it, I'm curious if we even need this defensive coding method at all?

dinahshi replied:
Preloader handles `nil` with a compact [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader.rb#L86). The [docs](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader.rb#L59-L62) say it should flatten nested arrays but now that I try it, I'm not seeing that behaviour. Will investigate further and see if either docs or code need to be changed.

I wanted to add this layer so that errors don't originate in Preloader, which isn't part of the public API. If we keep flatten, perhaps an addition to documentation (similar to the Preloader doc) would help explain the mutation to `records`.

dinahshi replied:
I checked on master and Preloader also doesn't flatten arrays there (i.e. passing a nested array into the `records` parameter causes an `undefined method for Array` error). I'll leave the nil check in, remove the flatten, and update the Preloader docs.

kaspth commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -678,6 +678,86 @@ def test_loading_with_one_association_with_non_preload
+  end
+
+  def test_load_associations_with_empty_array
+    assert_queries(0) do
```
Do we have an `assert_no_queries` assertion?

kaspth commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -678,6 +678,86 @@ def test_loading_with_one_association_with_non_preload
+
+  def test_load_associations_with_different_types
+    records = Author.all.to_a + Post.all.to_a
+    assert_queries(3) do
```
What's the third query for here?

dinahshi replied:
`Author has_many :comments, through: :posts` so, although the underlying association class `Comment` is the same for authors and posts, preloader fetches them separately. It first uses `Preloader::ThroughAssociation` to load author's comments (2 queries) and then `Preloader::Association` to load posts' comments (1 query). Relevant lines [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader.rb#L184). Here's what the queries look like:

```
SELECT "posts".* FROM "posts" WHERE "posts"."author_id" IN (?, ?, ?)
SELECT "comments".* FROM "comments" WHERE "comments"."post_id" IN (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
SELECT "comments".* FROM "comments" WHERE "comments"."post_id" IN (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?).
```

kaspth replied:
Ahh, I see. Thanks!

matthewd replied:
Maybe worth a `# posts->comments, authors->posts, authors->posts->comments`?

kaspth commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -678,6 +678,86 @@ def test_loading_with_one_association_with_non_preload
+      Author.load_associations(records, :comments)
+      records.each { |r| assert r.comments.loaded? }
+    end
+  end
```
Nice to know that we support polymorphic association loading as well!

I'm curious what happens if I keep the variables separate:

```ruby
authors, posts = Author.all.to_a, Post.all.to_a

ActiveRecord::Base.load_associations(authors + posts, :comments)

authors.map(&:comments) # Will SQL queries fire here? Or are the comments loaded?
```

dinahshi replied:
Comments are loaded, no SQL queries will fire.

kaspth replied:
👍, would it be clearer if we kept the variables separate in the tests and do `authors + posts`?

kaspth reviewed (commented):
Looking real good!

dinahshi commented:
I just realized that `ActiveRecord::Base.load_associations()` doesn't work at the moment. It throws `ActiveRecord::Base doesn't belong in a hierarchy descending from ActiveRecord` during scoping since it's currently set up in the Querying module using `delegate :load_associations, to: :all`.

I'm not sure how to rework this, would #load_associations have to be defined on the Core module as well as the Querying module? Is this effort worthwhile given that the method could be called on any subclass (i.e. do we want to support `ActiveRecord::Base.load_associations()` at all)?

cc/ @matthewd

matthewd commented:
> `ActiveRecord::Base doesn't belong in a hierarchy descending from ActiveRecord`

Well that's an impressively-confusing error message 😅 

Yes, let's move it out of Relation and into one of the model modules. At a glance it seems like a pretty equally awkward fit for both Querying and Associations, so Core looks to be the way to go. It's just a move, so there's no added complexity -- while I wouldn't encourage people to call it that way, all class methods are accessible via the relation automatically.

My thinking here is that 
1) it's conceptually not really something that acts on the current relation: the only existing methods it calls on `self` are `check_if_method_has_arguments!` and `build_preloader`, which are fine to inline and duplicate; 
2) it does seem useful for third party libraries to be able to `AR::Base.load_associations(..)` without needing to find a suitable subclass in the application; and 
3) if we were to fix the above error, it would complain that AR::Base is an abstract class [and therefore can't have a base class]... which would mean other user-defined abstract classes become invalid too (which would be totally correct, but more annoying when documenting valid callees for this method).

Commit:
Move load_associations from QueryMethods to Core

Commit:
Inline parameter checking

Commit:
Address comments on tests

Commit:
Update docs to reject nested arrays

Commit:
Remove load_associations delegation

dinahshi commented on activerecord/lib/active_record/associations/preloader/association.rb:
```diff
@@ -56,7 +56,7 @@ def owner_keys
           def owners_by_key
             unless defined?(@owners_by_key)
-              @owners_by_key = owners.each_with_object({}) do |owner, h|
+              @owners_by_key = owners.uniq.each_with_object({}) do |owner, h|
```
Rebased and looks like it's passing, thanks for pointing this out!

bogdan commented:
This PR has a lot of problems right now. Here is a brief list (some with failing tests):

* records with loaded associations may cause double loading (test 1)
* records with dirty associations will not properly merge in-memory and DB records (test 2) (see `CollectionAssociation#merge_target_lists`)
* through associations improperly uses loaded state which can have its own custom scope (test 3)
* through associations may override (with incomplete array because of through scope) and than reset the already loaded associations (no test invented yet)

My vision on fixing that is pretty blurry, but I can share these thoughts:
The handling of loaded and unloaded associations inside the same records array would need to improved. We can simply ensure association is not loaded before assigning association records to target inside `#associate_records_to_owner`. However this is ineffective due unnecessary loading of already loaded records. There should be a better way in generating separated preloader instances for records with loaded and unloaded associations. 

Also, through associations loading should never assign records to source and through reflection when the special scope is applied  (in terms of code - `Association#reset` should never be used).
It would require a serious refactoring of the preloader classes to ensure the code for  loading records and assigning them as association targets is separated and can be used independently.

Failing test cases (pluggable to `relations_test.rb`):

``` ruby
  def test_load_associations_ignore_loaded_records
    post1 = posts(:welcome)
    post2 = posts(:thinking)
    posts = [post1, post2]
    comment_ids = post2.comments.to_a.pluck(:id)
    Post.load_associations(posts, :comments)
    assert_equal comment_ids, post2.comments.to_a.pluck(:id)
  end

  def test_load_associations_into_dirty_association_merges_targets_properly
    post = posts(:welcome)
    ids = post.comment_ids
    post.comments << post.comments.first
    Post.load_associations(posts, :comments)
    assert_equal ids, post.comments.ids
  end

  require "models/department"
  require "models/hotel"
  require "models/chef"
  require "models/cake_designer"
  require "models/drink_designer"
  def test_load_associations_doesnt_reset_loaded_through_associations
    cake_designer = CakeDesigner.create!(chef: Chef.new)
    drink_designer = DrinkDesigner.create!(chef: Chef.new)
    chefs = [cake_designer.chef, drink_designer.chef]
    department = Department.create!(chefs: chefs)
    hotel = Hotel.create!(departments: [department])
    Hotel.load_associations([hotel], :cake_designers, :drink_designers)
    assert_equal chefs, hotel.chefs
    assert_equal [cake_designer], hotel.cake_designers
    assert_equal [drink_designer], hotel.drink_designers
  end
```

rafaelfranca commented:
Closing because the work is going to be done in effort to make preload public and we have a lot of conflicts.
